api/oneclick.py

```diff
@@ -0,0 +1,214 @@
+
+        error_response = self.check_and_log_errors(response=response, message=message, message_root="Checkout ")
+        if error_response:
+            return error_response
+
+        return resp_dict
+
+
+    def return_item(self, patron_id, item_id):
+        """
+        :param patron_id OneClick internal id
+        :param item_id isbn
+        """
+        resp_dict = self.checkout_item(patron_id, item_id, return_item=True)
+        if resp_dict == {}:
+            resp_dict = {'output':'SUCCESS'}
+        return resp_dict
+
+
+    def create_patron(self, patron_email=None, patron_cardno=None):
+        #http://api.oneclickdigital.us/v1/libraries/1931/patrons
+        '''
+        {
+        "libraryId":"1931",
+        "userName":"mickeymouse",
+        "password":"disney",
+        "email":"mickey@mouse.com",
+        "firstName":"Mic",
+        "lastName":"Mouse",
+        "libraryCardNumber":"13057226"
+        }
+        '''
+
+    def check_and_log_errors(self, response, message, message_root=""):
+        """ OneClick tries to communicate statuses and errors through http codes.
+        Malformed url requests will throw a 500, non-existent ids will get a 404, 
+        trying an action like checkout on a patron/item combo that's blocked 
+        (like if the item is already checked out, for example) will get a 409, etc..
+        Further details are usually elaborated on in the "message" field of the response.
+
+        :return True if non-fatal errors found, false otherwise (throws exceptions on bad errors).
+        """
+        if response.status_code != 200:
+            if not message:
+                message = response.text
+            self.log.warning("%s call failed: %s ", message_root, message)
+            return {"error_code":response.status_code, "message":message}
+        elif message:
+            # http code was OK, but info wasn't sucessfully read from db
+            # not exception-level fatal.  
+            self.log.warning("%s not retrieved: %s ", message_root, message)
+            return {"error_code":0, "message":message}
+
+        return None
+
+
+    def get_patron_internal_id(self, patron_email=None, patron_cardno=None):
+        """ Uses either an email address or a library card to identify a patron by.
+
+        :param patron_email 
+        :param patron_cardno
+        """
+        if patron_cardno: 
+            patron_identifier = patron_cardno
+        elif patron_email:
+            patron_identifier = patron_email
+        else:
+            # consider raising an exception, since we should call methods with nice arguments
+            return None
+
+        url = "%s/rpc/libraries/%s/patrons/%s" % (self.base_url, str(self.library_id), patron_identifier)
+
+        try:
+            response = self.request(url)
+        except Exception, e:
+            self.log.error("Patron id call failed: %r", e, exc_info=e)
+            raise ValueError(e.message)
```

Comment:
Yeah, OneClick handles a lot of responses through http errors.  Normally, an http request getting an exception means something dire has happened.  But with OneClick it could just mean a malformed identifier or a duplicate checkout request, etc..  Things that I'd put into the "oops" but not "stop the presses" category.  Don't know if ValueError is the precise right thing, but returning a None would make flow unintuitive, and leaving the original exception would cry "wolf" a bit much.